lib/symboldatabase.cpp

```diff
@@ -1425,6 +1428,49 @@ void SymbolDatabase::createSymbolDatabaseEscapeFunctions()
     }
 }
 
+void SymbolDatabase::createSymbolDatabaseExprIds()
+{
+    MathLib::bigint base = 0;
+    // Find highest varId
+    for (const Variable *var : mVariableList) {
+        if (!var)
+            continue;
+        base = std::max<MathLib::bigint>(base, var->declarationId());
+    }
+    for (const Scope * scope : functionScopes) {
+        MathLib::bigint id = base+1;
+        std::unordered_map<std::string, std::vector<Token*>> exprs;
+
+        // Assign IDs
+        for (Token* tok = const_cast<Token*>(scope->bodyStart); tok != scope->bodyEnd; tok = tok->next()) {
+            if (tok->varId() > 0) {
+                tok->exprId(tok->varId());
+            } else if (Token::Match(tok, "(|.|%cop%")) {
+                exprs[tok->str()].push_back(tok);
+                tok->exprId(id++);
+            }
+        }
+
+        // Apply CSE
+        for(auto&& p:exprs) {
+            std::vector<Token*> tokens = p.second;
```

Comment:
I guess a data copy of the vector is not needed  `const std::vector<Token *> &tokens = p.second;`. If you don't feel comfortable about the `const` because the items are changed, it is fine for me that it's non-const.